File: src/step-machine-public/args-massaging.ts
/**
 * step-machine-public — args massaging
 *
 * Pure JSONata-based mapping from logical args (the flow's flat context)
 * into a transport-specific shape. Lives here because:
 *
 *   - It is platform-free (only depends on jsonata).
 *   - Multiple adapters (Node spawn, HTTP, Azure Function) reuse it.
 *
 * Adapters call this helper as the first step inside `invoke(ref, input)`,
 * then perform their transport using `cmdArgs` / `body` / `url`.
 */

import { jsonata } from './jsonata-loader.js';
import type { ArgsMassaging } from '../cli/common/execution-interface.js';

export interface MassagedArgs {
  /** Resolved argv tail for local transports. */
  cmdArgs?: string[];
  /** Resolved request body object for http transports. */
  body?: unknown;
  /** Resolved final URL string for http transports. */
  url?: string;
}

/**
 * Evaluate `argsMassaging` against the supplied context.
 *
 * Throws with a step-name-tagged message if any expression fails — adapters
 * should catch and convert to a normalized failure result.
 */
export function resolveArgsMassaging(
  argsMassaging: ArgsMassaging | undefined,
  context: Record<string, unknown>,
  stepName: string,
): MassagedArgs {
  if (!argsMassaging || typeof argsMassaging !== 'object') return {};

  const out: MassagedArgs = {};

  if (Array.isArray(argsMassaging.cmdTemplate)) {
    const resolved: string[] = [];
    for (const expr of argsMassaging.cmdTemplate) {
      try {
        resolved.push(String(jsonata(expr).evaluate(context)));
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        throw new Error(
          `[step-machine-public] Step "${stepName}" argsMassaging.cmdTemplate failed on "${expr}": ${msg}`,
        );
      }
    }
    out.cmdArgs = resolved;
  }

  if (typeof argsMassaging.bodyTemplate === 'string') {
    try {
      out.body = jsonata(argsMassaging.bodyTemplate).evaluate(context);
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      throw new Error(
        `[step-machine-public] Step "${stepName}" argsMassaging.bodyTemplate failed: ${msg}`,
      );
    }
  }

  if (typeof argsMassaging.urlTemplate === 'string') {
    try {
      out.url = String(jsonata(argsMassaging.urlTemplate).evaluate(context));
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      throw new Error(
        `[step-machine-public] Step "${stepName}" argsMassaging.urlTemplate failed: ${msg}`,
      );
    }
  }

  return out;
}

File: standalone/core/pylib/cli/args_massaging.py
"""
args_massaging.py

JSONata-based mapping from logical args to transport-specific shape.

`argsMassaging` is a property of `ExecutionRef`, so honoring it is the job
of every adapter (subprocess, HTTP, Azure Function, etc.). This helper is
the shared pure-JSONata implementation reused by all adapters.

Mirror of src/cli/common/args-massaging.ts.
"""
from __future__ import annotations

from typing import Any, Dict, Optional

# Ensure vendored packages are importable
try:
    from ..vendor import __path__ as _vendor_path  # noqa: F401
except ImportError:
    pass

try:
    from jsonata import Jsonata as _Jsonata
    _JSONATA_AVAILABLE = True
except Exception:
    _JSONATA_AVAILABLE = False


def _jsonata_evaluate(expr: str, context: Dict[str, Any]) -> Any:
    if not _JSONATA_AVAILABLE:
        raise RuntimeError(
            "[args-massaging] jsonata package not installed. "
            "Run: python -m pip install -r standalone/requirements.txt"
        )
    return _Jsonata(expr).evaluate(context)


def resolve_args_massaging(
    args_massaging: Optional[Dict[str, Any]],
    context: Dict[str, Any],
    label: str,
) -> Dict[str, Any]:
    """
    Evaluate `argsMassaging` against the supplied context.

    Returns a dict with optional keys:
      cmdArgs: list[str]  — argv tail (local transports)
      body:    Any        — request body (HTTP) / stdin payload (local)
      url:     str        — final URL (HTTP transports)

    Raises ValueError with a label-tagged message if any expression fails.
    Adapters should catch and convert to a normalized failure result.
    """
    if not args_massaging or not isinstance(args_massaging, dict):
        return {}

    out: Dict[str, Any] = {}

    cmd_template = args_massaging.get("cmdTemplate")
    if isinstance(cmd_template, list):
        resolved = []
        for expr in cmd_template:
            try:
                resolved.append(str(_jsonata_evaluate(expr, context)))
            except Exception as ex:
                raise ValueError(
                    f'[{label}] argsMassaging.cmdTemplate failed on "{expr}": {ex}'
                ) from ex
        out["cmdArgs"] = resolved

    body_template = args_massaging.get("bodyTemplate")
    if isinstance(body_template, str):
        try:
            out["body"] = _jsonata_evaluate(body_template, context)
        except Exception as ex:
            raise ValueError(
                f'[{label}] argsMassaging.bodyTemplate failed: {ex}'
            ) from ex

    url_template = args_massaging.get("urlTemplate")
    if isinstance(url_template, str):
        try:
            out["url"] = str(_jsonata_evaluate(url_template, context))
        except Exception as ex:
            raise ValueError(
                f'[{label}] argsMassaging.urlTemplate failed: {ex}'
            ) from ex

    return out
